tools/generator/Java.Interop.Tools.Generator.Transformation/EnumMappings.cs

```diff
@@ -139,75 +141,22 @@ internal List<string> WriteEnumerations (string output_dir, Dictionary<string, E
-					string ns = enu.Key.Substring (0, enu.Key.LastIndexOf ('.')).Trim ();
-					string enoom = enu.Key.Substring (enu.Key.LastIndexOf ('.') + 1).Trim ();
-
-					sw.WriteLine ("namespace {0} {{", ns);
-					if (enu.Value.BitField)
-						sw.WriteLine ("  [System.Flags]");
-					sw.WriteLine ("  public enum {0} {{", enoom);
-
-					foreach (var member in enu.Value.Members) {
-						var managedMember = FindManagedMember (ns, enoom, enu.Value, member.Key, gens);
-						sw.WriteLine ("    [global::Android.Runtime.IntDefinition (" + (managedMember != null? "\"" + managedMember + "\"" : "null") + ", JniField = \"" + StripExtraInterfaceSpec (enu.Value.JniNames [member.Key]) + "\")]");
-						sw.WriteLine ("    {0} = {1},", member.Key.Trim (), member.Value.Trim ());
-					}
-					sw.WriteLine ("  }");
-					sw.WriteLine ("}");
-				}
-			}
-			return files;
-		}
-
-		string StripExtraInterfaceSpec (string jniFieldSpec)
-		{
-			return jniFieldSpec.StartsWith ("I:", StringComparison.Ordinal) ? jniFieldSpec.Substring (2) : jniFieldSpec;
-		}
-
-		string FindManagedMember (string ns, string enumName, EnumDescription desc, string enumFieldName, IEnumerable<GenBase> gens)
-		{
-			if (desc.FieldsRemoved)
-				return null;
-
-			var jniMember = desc.JniNames [enumFieldName];
-			if (string.IsNullOrWhiteSpace (jniMember)) {
-				// enum values like "None" falls here.
-				return null;
-			}
-			return FindManagedMember (jniMember, gens);
-		}
+			var files = new ConcurrentBag<string> ();
 
-		WeakReference cache_found_class;
+			Parallel.ForEach (enums, enu => {
+				var path = Path.Combine (output_dir, GetFileName (enu.Key, useShortFileNames) + ".cs");
+				files.Add (path);
 
-		string FindManagedMember (string jniMember, IEnumerable<GenBase> gens)
-		{
-			string package, type, member;
-			ParseJniMember (jniMember, out package, out type, out member);
-			var fullJavaType = (string.IsNullOrEmpty (package) ? "" : package + ".") + type;
-
-			var cls = cache_found_class != null ? cache_found_class.Target as GenBase : null;
-			if (cls == null || cls.JniName != fullJavaType) {
-				cls = gens.FirstOrDefault (g => g.JavaName == fullJavaType);
-				if (cls == null) {
-					// The class was not found e.g. removed by metadata fixup.
-					return null;
+				using (var sw = File.CreateText (path)) {
+					var generator = new EnumGenerator (sw);
+					generator.WriteEnumeration (enu, gens);
 				}
-				cache_found_class = new WeakReference (cls);
-			}
-			var fld = cls.Fields.FirstOrDefault (f => f.JavaName == member);
-			if (fld == null) {
-				// The field was not found e.g. removed by metadata fixup.
-                                return null;
-			}
-			return cls.FullName + "." + fld.Name;
+			});
+
+			return files.ToList ();
 		}
 
-		Dictionary<string,string> file_name_map = new Dictionary<string, string> ();
+		ConcurrentDictionary<string,string> file_name_map = new ConcurrentDictionary<string, string> ();
```

Comment:
If `file_name_map` is now a `ConcurrentDictionary`, shouldn't `GetFileName()` be updated to take advantage of that?  The current implementation isn't using `file_name_map` in a thread-safe manner:

```csharp
			if (file_name_map.TryGetValue (file, out s))
				return s;
			s = file_name_map.Count.ToString ();
			file_name_map [file] = s;
```

This should probably use [`ConcurrentDictionary.GetOrAdd()`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netframework-4.8):

```csharp
string GetFileName (string file, bool useShortFileNames)
{
	if (!useShortFileNames)
		return file;

	return file_name_map.GetOrAdd (file, (key, arg) => arg.ToString (), file_name_map.Count);
}
```

I'm not *entirely* convinced that this would be thread safe, due to the `file_name_map.Count` use, but the current implementation isn't at all thread safe.

Given that I'm not convinced that even the above can be made to be thread-safe, I *suspect* that the better course of action is to *continue* using a `Dictionary<string, string>` and instead add appropriate `lock(file_name_map)` calls, so that we know it's in a consistent state:

```csharp
			if (file_name_map.TryGetValue (file, out s))
				return s;
			lock (file_name_map) {
				s = file_name_map.Count.ToString ();
				file_name_map [file] = s;
			}
```

Reply:
Good catch, went with the `lock`.